NBitcoin/Utils/ThreadSafeList.cs

```diff
@@ -92,12 +96,14 @@ public T FindOrCreate<U>() where U : T, new()
 
 		public IEnumerator<T> GetEnumerator()
 		{
-			List<T> list = null;
-			lock (_lock)
+			if (_EnumeratorList == null)
 			{
-				list = _Behaviors.ToList();
+				lock (_lock)
+				{
+					_EnumeratorList = _Behaviors.ToList();
+				}
 			}
-			return list?.GetEnumerator();
+			return _EnumeratorList?.GetEnumerator();
```

Comment:
Please capture `enumeratorList` in a lock variable so we are sure GetEnumerator never returns null.